Allow selection of target agent

matelakat opened the pull request:
Enable users to specify explicitly a target agent for router migration either by using an agent id or a host id

Commit:
Implement host and agent id based agent pickers

To support selecting explicitly a destination agent for router
migration, two new agent picking strategies are added, one for
selecting the destination based on the agent's id, the other to select
the agent based on it's host.

matelakat commented:
![](https://github.trello.services/images/mini-trello-icon.png) [neutron-ha-tool: enable users to specify target agent](https://trello.com/c/gPIctZBJ/8-neutron-ha-tool-enable-users-to-specify-target-agent)

aplanas commented on files/default/neutron-ha-tool.py:
```diff
@@ -233,6 +236,10 @@ def run(args):
     else:
         raise ValueError('Invalid agent_selection_mode')
 
+    if args.target_agent is not None:
```
Comparison with None is very unusual (only in lxml and because a bug). `if args.target_agent:` is mostly expected, or `args.target_agent` can be 0, '' or something like that?

matelakat replied:
You are right, I think we can assume that having a falsey value is an error here, so we can get rid of the ugly `is not` part.

aplanas commented on files/default/neutron-ha-tool.py:
```diff
@@ -233,6 +236,10 @@ def run(args):
 
+    if args.target_agent is not None:
+        Configuration.agent_picker_class = SingleAgentPicker
+        SingleAgentPicker.agent_selection_value = args.target_agent
```
uhm this is a bit ugly, this value is now for all SingleAgentPicker instances, because you are updating the class

But TBH I do not see a better solution if we work with classes instead of instances. 

Maybe a better solution is to pass the strategy into `migrate_l3_routers_from_agent` via `l3_agent_evacuate` and `l3_agent_migrate` as a parameter. But I agree that this will require more refactoring : (((

matelakat replied:
Yes, I feel your pain. Modifying class variables is sub-optimal. What we could do is to change what is being stored in `agent_picker_class`, and let's say we store an `agent_picker` - which will be an instance. In that case we would need to add one further parameter to the agent picker interface, something like `set_agents`. Let me add a commit, and see if that design would make it any better.

aspiers replied:
Agreed - this is related to [what I wrote 2 days ago](https://github.com/SUSE-Cloud/cookbook-openstack-network/pull/9#discussion_r94273578), before I noticed these older comments.

matelakat commented:
@aplanas I appended a commit to the end of the PR, which refactors the pickers a bit - I'm interested in your thoughts!

aplanas reviewed (approved):
Thanks for the refactoring. IMO is much better like this.

vuntz reviewed (approved):
LGTM

vuntz commented on files/default/neutron-ha-tool.py:
```diff
@@ -964,6 +964,27 @@ def pick(self):
+            agent_values = [agent.get(key) for key in self.agent_keys]
+            if value in agent_values:
+                return agent
+        raise IndexError('Cannot find desired agent')
```
Maybe give the id of the agent in the error?

aspiers replied:
Yes please, otherwise Murphy's Law dictates we'll inevitably see this and end up in a situation which is impossible to debug.

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -963,6 +963,27 @@ def pick(self):
+    def __init__(self, qclient, agents):
+        self.agents = agents
+
+    def find_agent(self, value):
```
`value` is a very vague variable name.  Please give it a name which reflects what the value actually represents.  IIUC, it actually represents `agent_id_or_host_id`, and IMHO the ugliness of that name is a code smell which suggests it's not ideal to overload a single method with two different types of search.  For example there is perhaps a chance (however astronomically small) that a given agent id is also a valid host id.  Perhaps a more likely issue would be that someone in the future writes some buggy code which accidentally passes an agent id when they actually meant to pass a host id, or vice-versa.  This overloading would hide that bug.  Or even more likely, someone manually passes the wrong id via the CLI.

I'd prefer to have separate code paths, e.g. `find_agent_by_id` and `find_agent_by_host_id`, or at least `def find_agent(self, object_type, object_id)`.  This would also make it clearer how often each of the code paths get used.  If one never gets used then it would be better to drop it.

matelakat replied:
Right, I guess the best would be to enable users to specify explicitly what the string refers to.

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -240,15 +240,14 @@ def run(args):
 
     if args.agent_selection_mode == 'random':
-        Configuration.agent_picker_class = RandomAgentPicker
+        Configuration.agent_picker = RandomAgentPicker()
```
Yes, this is much better.  As [previously mentioned](https://github.com/SUSE-Cloud/cookbook-openstack-network/pull/9#discussion_r94273578) there's still a code smell since you're using a class variable to store the instance of the strategy subclass.  I would prefer it if you instantiated from the `Configuration` class and then passed that configuration object around to whatever needs it.  That might even result in showing that a separate configuration class is not needed, since the only configuration currently is which strategy subclass instance needs to be used.

BTW I can live with it, but I still wanted to be annoying and point out that this commit goes directly against [part of my advice for how to amend PRs](http://blog.adamspiers.org/2015/03/24/why-and-how-to-correctly-amend-github-pull-requests/#sec-2).

aspiers reviewed (changes requested):
Hmm, it seems I am the annoying one (yet again).

Most of this is fantastic.  If you could just include the agent id in the error and improve that variable name then I'm OK if you ignore my other nitpicks :-)

matelakat commented:
Let me fix both issues.

Commit:
Enable users to explicitly specify target agent

To support selection of a destination agent for router migration,
user can provide either a host or an agent id - these values are
used to create the appropriate agent picking strategies.

matelakat commented:
I think I made the change better by separating the two agent pickers based on what the user wanted - this removes the issues, and I think it's a better design. Please note that I could have extracted some common base class for the agent pickers, but I did not do that now - we could do that as a next step. For the two new classes it wouldn't have made much of a difference in terms of lines of code if I extracted them.

vuntz commented on files/default/neutron-ha-tool.py:
```diff
@@ -968,6 +979,40 @@ def pick(self):
+
+
+class HostIdBasedAgentPicker(object):
+    def __init__(self, host_id):
```
`host` would be better than `host_id`, but no big deal ;-)

vuntz commented:
Needs another rebase, but I guess you already know that...

matelakat commented:
Yep, doing that now.

matelakat commented:
@vuntz @aspiers - this has been amended, please give it another review.

matelakat commented:
@rhafer Given that the change has been amended, would you mind taking another look please?

rhafer reviewed (approved):
Have you considered making ```--target-agent-id```, ```--targed-host``` and ```--agent-selection-mode``` mutually exclusive? (should be pretty simple using ```ArgumentParser.add_mutually_exclusive_group()```

That being said I am still fine merging this as is.

matelakat commented:
> Have you considered making --target-agent-id, --targed-host and --agent-selection-mode mutually exclusive? (should be pretty simple using ArgumentParser.add_mutually_exclusive_group()
> 
> That being said I am still fine merging this as is.

Are you happy having that as a separate PR?

rhafer commented:
> Are you happy having that as a separate PR?

Sure

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -117,6 +117,12 @@ def parse_args():
+    ap.add_argument('--target-agent-id', default=None,
+                    help='Explicitly select a target agent by specifying an '
+                         'agent id.')
+    ap.add_argument('--target-host', default=None,
```
Shouldn't these options be mutually exclusive?

aspiers replied:
Oh, @rhafer already pointed that out ... never mind.

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -248,6 +254,11 @@ def run(args):
 
+    if args.target_agent_id:
+        agent_picker = AgentIdBasedAgentPicker(args.target_agent_id)
+    if args.target_host:
```
If both are specified, `--target-agent-id` will be ignored, which is unintuitive.  If you `add_mutually_exclusive_group` above then that defends against this, but it would still be better to change this to `elif`.

aspiers reviewed (changes requested):
Almost perfect now ;-)

aspiers commented:
I'm OK with the mutually exclusive thing being done as a follow-up, although wouldn't it have been just as easy to fix it here?

Also, #toomanyreviewers ...
